File: scripts/check-ci-scope.mjs
import assert from 'node:assert/strict'
import {spawnSync} from 'node:child_process'
import {readFileSync} from 'node:fs'
import {dirname, resolve} from 'node:path'
import {fileURLToPath} from 'node:url'

const root = resolve(dirname(fileURLToPath(import.meta.url)), '..')
const jobs = ['plugin', 'runtime', 'eval', 'issue', 'omp']

function verifyAggregate(env) {
  if (env.DETECTION !== 'success' || env.REPOSITORY !== 'success') return false
  for (const name of jobs) {
    const selectedValue = env[`${name.toUpperCase()}_SELECTED`]
    const result = env[`${name.toUpperCase()}_RESULT`]
    if (!['true', 'false'].includes(selectedValue)) return false
    if (selectedValue === 'true' && result !== 'success') return false
    if (selectedValue === 'false' && result !== 'skipped') return false
  }
  return true
}

if (process.argv[2] === '--aggregate') {
  if (!verifyAggregate(process.env)) {
    console.error('CI Success rejected a failed, skipped, cancelled, or invalid required job')
    process.exitCode = 1
  }
} else {
  const [{default: yaml}, {default: picomatch}] = await Promise.all([
    import('js-yaml'),
    import('picomatch'),
  ])
  const workflowText = readFileSync(resolve(root, '.github/workflows/ci.yml'), 'utf8')
  const workflow = yaml.load(workflowText)
  const filters = yaml.load(readFileSync(resolve(root, '.github/ci-filters.yml'), 'utf8'))

  function filterMatches(patterns, paths) {
    const include = patterns.filter(pattern => !pattern.startsWith('!'))
    const exclude = patterns.filter(pattern => pattern.startsWith('!')).map(pattern => pattern.slice(1))
    return paths.some(path => include.some(pattern => picomatch(pattern, {dot: true})(path)) &&
      !exclude.some(pattern => picomatch(pattern, {dot: true})(path)))
  }

  function selected(paths) {
    const matched = Object.fromEntries(Object.entries(filters).map(([name, patterns]) =>
      [name, filterMatches(patterns, paths)]))
    return Object.fromEntries(jobs.map(job => [job,
      matched.shared || matched.unknown || matched[job]]))
  }

  // The pinned action reports a rename as both the old and new path.
  const fixtures = [
    {name: 'config-only', paths: ['.github/workflows/ci.yml'], expected: jobs},
    {name: 'plugin component', paths: ['.claude-plugin/plugin.json'], expected: ['plugin', 'omp']},
    {name: 'runtime component', paths: ['hooks/check.py'], expected: ['runtime']},
    {name: 'eval component', paths: ['omp_configs/personal.yml'], expected: ['eval']},
    {name: 'issue component', paths: ['infra/github/labels.tf'], expected: ['issue']},
    {name: 'Omp component', paths: ['extensions/daily-driver.js'], expected: ['runtime', 'omp']},
    {name: 'mixed components', paths: ['agents/reviewer.md', 'infra/github/labels.tf'], expected: ['plugin', 'issue']},
    {name: 'shared script', paths: ['scripts/new-validation.py'], expected: jobs},
    {name: 'unknown path', paths: ['new-component/source.txt'], expected: jobs},
    {name: 'unknown root file', paths: ['new-root-file.txt'], expected: jobs},
    {name: 'deleted component file', paths: ['hooks/removed.py'], expected: ['runtime']},
    {name: 'cross-component rename', paths: ['omp_configs/old.yml', 'skills/new-skill/SKILL.md'], expected: ['plugin', 'runtime', 'eval', 'issue', 'omp']},
    {name: 'new model overlay', paths: ['omp_configs/new-model.yml'], expected: ['eval']},
  ]
  for (const fixture of fixtures) {
    const actual = selected(fixture.paths)
    const got = Object.keys(actual).filter(job => actual[job])
    assert.deepEqual(got, fixture.expected, fixture.name)
  }
  assert.equal(workflow.jobs.changes.permissions['pull-requests'], 'read')
  assert.equal(workflow.jobs['ci-success'].if, 'always()')
  assert.equal(workflow.jobs['ci-success'].name, 'CI Success')
  assert.deepEqual(workflow.jobs['ci-success'].needs, [
    'changes', 'repository-checks', 'plugin-validity', 'runtime',
    'eval-tooling', 'issue-infra', 'omp-integration',
  ])
  assert.match(workflowText, /dorny\/paths-filter@ceb8a2b8f2d89434be7ff52d3de7ec3738c5cc9d/)
  assert.match(workflowText, /predicate-quantifier: some-with-excludes/)

  function aggregateCase(name, values, expected) {
    const result = spawnSync(process.execPath, [fileURLToPath(import.meta.url), '--aggregate'], {
      cwd: root,
      env: {...process.env, ...values},
      encoding: 'utf8',
    })
    assert.equal(result.status === 0, expected, `aggregate case ${name}: ${result.stderr}`)
  }
  const valid = {DETECTION: 'success', REPOSITORY: 'success'}
  for (const job of jobs) {
    valid[`${job.toUpperCase()}_SELECTED`] = job === 'eval' ? 'true' : 'false'
    valid[`${job.toUpperCase()}_RESULT`] = job === 'eval' ? 'success' : 'skipped'
  }
  aggregateCase('intentional skips', valid, true)
  for (const [name, selectedValue, resultValue] of [
    ['selected failure', 'true', 'failure'],
    ['selected skip', 'true', 'skipped'],
    ['selected cancellation', 'true', 'cancelled'],
    ['unselected run', 'false', 'success'],
  ]) {
    aggregateCase(name, {...valid, PLUGIN_SELECTED: selectedValue, PLUGIN_RESULT: resultValue}, false)
  }
  aggregateCase('detection failure', {...valid, DETECTION: 'failure'}, false)
  aggregateCase('unconditional failure', {...valid, REPOSITORY: 'failure'}, false)
  aggregateCase('missing selection output', {...valid, EVAL_SELECTED: ''}, false)
  console.log('check-ci-scope: paths-filter fixtures and CI Success contract passed')
}

File: scripts/check-ci-scope.py
#!/usr/bin/env python3
"""Assert CI scope classification and required-check aggregation offline."""
from __future__ import annotations

import importlib.util
import os
import subprocess
import tempfile
from pathlib import Path
from unittest.mock import patch

SCRIPT = Path(__file__).with_name("ci-scope.py")
spec = importlib.util.spec_from_file_location("ci_scope", SCRIPT)
assert spec and spec.loader
ci_scope = importlib.util.module_from_spec(spec)
spec.loader.exec_module(ci_scope)


def require(condition: bool, message: str) -> None:
    """Fail loudly when an offline CI contract does not hold."""
    if not condition:
        raise AssertionError(message)


with tempfile.TemporaryDirectory() as directory:
    root = Path(directory)
    experiments = root / "evals" / "experiments"
    experiments.mkdir(parents=True)
    with patch.object(ci_scope, "ROOT", root):
        require(ci_scope.classify(["omp_configs/personal.yml"]) == set(), "personal overlay without an experiment stays skipped")
        (experiments / "classes-personal.yaml").touch()
        require(ci_scope.classify(["omp_configs/personal.yml"]) == {"eval"}, "overlay with a suite selects eval validation")
require(ci_scope.classify([".claude-plugin/plugin.json"]) == {"plugin"}, "manifest changes must select plugin validation")
require(ci_scope.classify(["skills/pr/SKILL.md"]) == {"plugin", "runtime"}, "skill changes must select manifest and route validation")
require(ci_scope.classify(["evals/tasks/example.yaml"]) == {"eval", "runtime"}, "eval changes must run eval and shared runtime validators")
require(ci_scope.classify(["infra/github/labels.tf"]) == {"issue"}, "infrastructure changes must select issue validation")
require(ci_scope.classify(["unknown/new-file.dat"]) == set(ci_scope.GROUPS), "unknown files must fail open")
with patch.object(ci_scope, "changed_paths", side_effect=OSError("missing base")):
    require(
        ci_scope.scope_for("missing", "head") == set(ci_scope.GROUPS),
        "unavailable diff base must fail open",
    )
with patch.object(ci_scope, "changed_paths", return_value=["evals/tasks/example.yaml"]) as diff:
    require(ci_scope.scope_for("before", "head", False) == {"eval", "runtime"}, "pushes must classify before/after changes")
    diff.assert_called_once_with("before", "head", False)
with tempfile.TemporaryDirectory() as directory:
    root = Path(directory)
    subprocess.run(["git", "init", "-q", str(root)], check=True)
    overlays = root / "omp_configs"
    overlays.mkdir()
    (overlays / "old.yml").write_text("modelRoles:\n  task: example/model\n")
    subprocess.run(["git", "-C", str(root), "add", "omp_configs/old.yml"], check=True)
    subprocess.run(
        ["git", "-C", str(root), "-c", "user.name=CI", "-c", "user.email=ci@example.invalid",
         "commit", "-qm", "initial overlay"],
        check=True,
    )
    before = subprocess.check_output(["git", "-C", str(root), "rev-parse", "HEAD"], text=True).strip()
    subprocess.run(["git", "-C", str(root), "mv", "omp_configs/old.yml", "omp_configs/new.yml"], check=True)
    subprocess.run(
        ["git", "-C", str(root), "-c", "user.name=CI", "-c", "user.email=ci@example.invalid",
         "commit", "-qm", "rename overlay"],
        check=True,
    )
    after = subprocess.check_output(["git", "-C", str(root), "rev-parse", "HEAD"], text=True).strip()
    previous = Path.cwd()
    try:
        os.chdir(root)
        paths = ci_scope.changed_paths(before, after, False)
    finally:
        os.chdir(previous)
    experiments = root / "evals" / "experiments"
    experiments.mkdir(parents=True)
    (experiments / "classes-old.yaml").touch()
    with patch.object(ci_scope, "ROOT", root):
        require(ci_scope.classify(paths) == {"eval"}, "renaming an overlay must detect its stale experiment")
    require(set(paths) == {"omp_configs/old.yml", "omp_configs/new.yml"}, "renames must retain both changed input paths")

# Exercise the exact shell block CI runs, not a second aggregate implementation.
workflow = (SCRIPT.parent.parent / ".github/workflows/ci.yml").read_text()
aggregate_step = workflow.split("      - name: Verify every job succeeded\n", 1)[1]
run_block = aggregate_step.split("        run: |\n", 1)[1]
commands = []
for line in run_block.splitlines():
    if not line.startswith("          "):
        break
    commands.append(line[10:])
require(bool(commands), "CI Success must contain an executable aggregation step")
for results, expected in (("success success", True), ("success failure", False),
                          ("success cancelled", False), ("", False)):
    completed = subprocess.run(
        ["bash", "-c", "\n".join(commands)], env={"RESULTS": results},
        capture_output=True, text=True,
    )
    require((completed.returncode == 0) == expected, f"CI Success aggregation mishandles {results!r}")
print("check-ci-scope: classification and required-status assertions passed")
